test-data/unit/check-enum.test

```diff
@@ -854,3 +902,81 @@ def process(response: Union[str, Reason] = '') -> str:
         return 'PROCESSED: ' + response
 
 [builtins fixtures/primitives.pyi]
+
+[case testEnumReachabilityDisabledGivenCustomEquality]
+from typing import Union
+from enum import Enum
+
+class Parent(Enum):
+    def __ne__(self, other: object) -> bool: return True
+
+class Foo(Enum):
+    A = 1
+    B = 2
+    def __eq__(self, other: object) -> bool: return True
+
+class Bar(Parent):
+    A = 1
+    B = 2
+
+class Ok(Enum):
+    A = 1
+    B = 2
+
+x: Foo
+if x is Foo.A:
+    reveal_type(x)  # N: Revealed type is 'Literal[__main__.Foo.A]'
+else:
+    reveal_type(x)  # N: Revealed type is 'Literal[__main__.Foo.B]'
+
+if x == Foo.A:
+    reveal_type(x)  # N: Revealed type is '__main__.Foo'
+else:
+    reveal_type(x)  # N: Revealed type is '__main__.Foo'
+
+y: Bar
+if y is Bar.A:
+    reveal_type(y)  # N: Revealed type is 'Literal[__main__.Bar.A]'
+else:
+    reveal_type(y)  # N: Revealed type is 'Literal[__main__.Bar.B]'
+
+if y == Bar.A:
+    reveal_type(y)  # N: Revealed type is '__main__.Bar'
+else:
+    reveal_type(y)  # N: Revealed type is '__main__.Bar'
+
+z1: Union[Bar, Ok]
+if z1 is Ok.A:
+    reveal_type(z1)  # N: Revealed type is 'Literal[__main__.Ok.A]'
+else:
+    reveal_type(z1)  # N: Revealed type is 'Union[__main__.Bar, Literal[__main__.Ok.B]]'
+
+z2: Union[Bar, Ok]
+if z2 == Ok.A:
+    reveal_type(z2)  # N: Revealed type is 'Union[__main__.Bar, __main__.Ok]'
+else:
+    reveal_type(z2)  # N: Revealed type is 'Union[__main__.Bar, __main__.Ok]'
+[builtins fixtures/primitives.pyi]
+
+[case testEnumReachabilityWithChaining]
+from enum import Enum
+class Foo(Enum):
+    A = 1
+    B = 2
+
+x: Foo
+y: Foo
+if x is Foo.A is y:
+    reveal_type(x)  # N: Revealed type is 'Literal[__main__.Foo.A]'
+    reveal_type(y)  # N: Revealed type is 'Literal[__main__.Foo.A]'
+else:
+    reveal_type(x)  # N: Revealed type is '__main__.Foo'
+    reveal_type(y)  # N: Revealed type is '__main__.Foo'
+
+if x == Foo.A == y:
```

Comment:
Does x == y == Foo.A work? (It feels like we ought to support this or I'm not sure how much this helps?)